File: widgets/symphony_list/src/widget.ts
// For licensing see accompanying LICENSE file.
// Copyright (C) 2023 Apple Inc. All Rights Reserved.

import { MODULE_NAME, MODULE_VERSION } from './version';
import {
  DOMWidgetModel,
  DOMWidgetView,
  ISerializers,
} from '@jupyter-widgets/base';

import {serializers, JupyterWidget } from '@apple/symphony-lib';

import SymphonyList from './SymphonyList.svelte';

export class SymphonyListModel extends DOMWidgetModel {
  defaults() {
    return {
      ...super.defaults(),
      _model_name: SymphonyListModel.model_name,
      _model_module: SymphonyListModel.model_module,
      _model_module_version: SymphonyListModel.model_module_version,
      _view_name: SymphonyListModel.view_name,
      _view_module: SymphonyListModel.view_module,
      _view_module_version: SymphonyListModel.view_module_version,
    };
  }
  static serializers: ISerializers = serializers;

  static model_name = 'SymphonyListModel';
  static model_module = MODULE_NAME;
  static model_module_version = MODULE_VERSION;
  static view_name = 'SymphonyListView'; // Set to null if no view
  static view_module = MODULE_NAME; // Set to null if no view
  static view_module_version = MODULE_VERSION;
}

export class SymphonyListView extends DOMWidgetView {
  render() {
    new JupyterWidget({
      target: this.el,
      props: {
        model: this.model,
        widget: SymphonyList,
      },
    });
  }
}

File: widgets/symphony_list/symphony_list/widget.py
# For licensing see accompanying LICENSE file.
# Copyright (C) 2023 Apple Inc. All Rights Reserved.

from pathlib import Path

import symphony_ui
from traitlets import Unicode

from ._frontend import module_name, module_version


class SymphonyList(symphony_ui.SymphonyWidget):
    _model_name = Unicode(
        'SymphonyListModel').tag(sync=True)
    _view_name = Unicode(
        'SymphonyListView').tag(sync=True)
    _model_module = Unicode(module_name).tag(sync=True)
    _model_module_version = Unicode(module_version).tag(sync=True)
    _view_module = Unicode(module_name).tag(sync=True)
    _view_module_version = Unicode(module_version).tag(sync=True)
    name = 'SymphonyList'
    description = "A Symphony component that displays a view of data instances"

    def __init__(self,
                 width: str = 'XXL',
                 height: str = 'M',
                 page: str = 'List',
                 **kwargs
                 ):
        """A Symphony component that displays a view of data instances

        Parameters
        ----------
        width : str, optional
            By default "XXL".
        height : str, optional
            By default "M".
        page : str, optional
            Which page of the dashboard to show the widget on, by default "SymphonyList"

        Returns
        -------
        SymphonyList
        """
        super().__init__(**kwargs)
        self.width: str = width
        self.height: str = height
        self.page: str = page
        self.spec = symphony_ui.WidgetSpec(
            width=width,
            height=height,
            page=page,
            name=self.name,
            description=self.description
        )

    def js_path(self):
        app_files_path = Path(
            (Path(__file__).parent), 'standalone', 'widgets')
        js_path = Path(app_files_path, 'SymphonyList.js')
        map_path = Path(
            app_files_path, 'SymphonyList.js.map')
        return js_path, map_path
